src/ocean/task/util/Timer.d

```diff
@@ -124,6 +124,113 @@ unittest
 
+    Similar to `theScheduler.await` but also has waiting timeout. Calling task
+    will be resumed either if awaited task finished or timeout is hit, whichever
+    happens first.
+
+    Params:
+        task = task to await for
+        micro_seconds = timeout duration
+
+    Returns:
+        'true' if resumed via timeout, 'false' otherwise
+
+*******************************************************************************/
+
+public bool awaitOrTimeout ( Task task, uint micro_seconds )
+{
+    auto context = Task.getThis();
+    assert (context !is null);
+
+    if (.timer is null)
+        .timer = new typeof(timer);
+
+    auto scheduled_event = .timer.schedule(
+        // EventData setup is run from the same fiber so it is ok to reference
+        // variable from this function stack
+        ( ref EventData event )
+        {
+            event.to_resume = context;
+        },
+        // Callback of fired timer is run from epoll context and here it is
+        // only legal to use data captured as EventData field (or other heap
+        // allocated data)
+        ( ref EventData event )
+        {
+            debug_trace("Resuming task <{}> because of await timeout",
+                cast(void*) event.to_resume);
+            event.to_resume.resume();
+        },
+        micro_seconds
+    );
+
+    task.terminationHook(&scheduled_event.unregister);
+    task.terminationHook(&context.resume);
+
+    // force async scheduling to avoid checking if this context needs
+    // suspend/resume and do it unconditionally
+    theScheduler.queue(task);
+    context.suspend();
+
+    if (task.used())
+    {
+        // resumed because of timeout
+        task.removeTerminationHook(&context.resume);
+        task.removeTerminationHook(&scheduled_event.unregister);
+        return true;
+    }
+    else
+    {
+        // resumed because awaited task has finished
+        // timer will be unregistered by its termination hook
+        return false;
+    }
+}
+
+///
+unittest
+{
+    initScheduler(SchedulerConfiguration.init);
+
+    .timer = new typeof(timer);
+
+    class InfiniteTask : Task
+    {
+        override public void run ( )
+        {
+            for (;;) .wait(100);
+        }
+    }
+
+    class RootTask : Task
+    {
+        Task to_wait_for;
+
+        override public void run ( )
+        {
+            bool timeout = .awaitOrTimeout(this.to_wait_for, 200);
+            test(timeout);
+
+            // `awaitOrTimeout` itself won't terminate awaited task on timeout,
+            // it will only "detach" it from the current context. If former is
+            /// desired, it can be trivially done at the call site:
+            if (timeout)
+                this.to_wait_for.kill();
+        }
+    }
+
+    auto root = new RootTask;
+    root.to_wait_for = new InfiniteTask;
+
+    theScheduler.schedule(root);
+    theScheduler.eventLoop();
+
+    test(!root.used());
+    test(!root.to_wait_for.used());
```

Comment:
Any way to write a quick test with waiting for the task that actually finishes?

Reply:
(with no timeout)

Reply:
Sure, will do (it would be a copy of one for `await` pretty much)

Reply:
Good catch, I actually had one bug there!
Updated.